Updates for purchase prompt

ZoteTheMighty opened the pull request:
Adds a bunch of service stubs and enums that are needed for testing purchase prompt.

LPGhatguy commented on lib/Enum/KeyCode.lua:
```diff
@@ -0,0 +1,258 @@
+	World92 = 252,
+	World93 = 253,
+	World94 = 254,
+	World95 = 255,
```
what on earth are these keys

ZoteTheMighty replied:
:/

Kampfkarren replied:
The only reference I can find to them is [in here](http://sdl.beuc.net/sdl.wiki/SDLKey), something about different keyboard layouts??? Doubt they do anything in Roblox though.

LPGhatguy commented on lib/instances/GuiService.lua:
```diff
@@ -25,4 +27,13 @@ end
 end
 
+function GuiService.prototype:IsTenFootInterface()
+end
```
This should probably return a bool like the real `IsTenFootInterface` does; we can just return `false` here until we have configurable habitats.

LPGhatguy commented on lib/instances/GuiService_spec.lua:
```diff
@@ -30,4 +31,18 @@ describe("instances.GuiService", function()
+
+	describe("IsTenFootInterface", function()
+		it("should exist", function()
+			GuiService:new():IsTenFootInterface()
```
Let's assert that it returns false too!

LPGhatguy commented on lib/instances/HttpService.lua:
```diff
@@ -11,4 +12,21 @@ function HttpService.prototype:JSONDecode(input)
+	local argType = typeof(wrapInCurlyBraces)
+	if argType == "userdata" or argType == "table" or argType == "function" then
+		error("Unable to cast argument to bool", 2)
+	end
```
Let's be more strict here and only accept a `boolean`.

LPGhatguy commented on lib/instances/ScreenGui.lua:
```diff
@@ -33,6 +33,34 @@ ScreenGui.properties.DisplayOrder = InstanceProperty.normal({
+	getDefault = function()
+		return false
+	end,
+})
```
I think there's a `typed` utility we can use instead of checking things manually here.

LPGhatguy commented on lib/instances/Workspace.lua:
```diff
@@ -13,4 +13,6 @@ end
 
 Workspace.properties.CurrentCamera = InstanceProperty.normal({})
 
+Workspace.properties.DistributedGameTime = InstanceProperty.normal({})
```
This probably needs a default value of `0` and probably made read-only

Kampfkarren replied:
Just a fun fact, DistributedGameTime isn't actually read only, but it's changed back to the real value immediately afterwards. Roblox issue that Lemur shouldn't emulate.

ZoteTheMighty commented:
PR comments, plus:
- Made some additional changes to `ScreenGui.DisplayOrder` to make it consistent with the fields I added.
- Removed overzealous tests for `ScreenGui` properties since I'm using `InstanceProperty.typed` now
- Made `ScreenGui.AutoLocalize` return true by default, which is what it does in Roblox

LPGhatguy commented on lib/instances/ScreenGui.lua:
```diff
@@ -19,17 +19,21 @@ ScreenGui.properties.AbsoluteSize = InstanceProperty.readOnly({
+	getDefault = function()
+		return true
 	end,
+})
```
I guess we don't have `GuiBase` or `GuiBase2d` yet, but this property really belongs there.

ZoteTheMighty replied:
Oh yeah, I didn't realize that. Should I make GuiBase2d and pull it out?

ZoteTheMighty replied:
Actually, `AbsoluteSize` and `AbsolutePosition` are from `GuiBase2d` as well. And `ZIndexBehavior` is from `LayerCollector`...

LPGhatguy commented on lib/instances/ScreenGui_spec.lua:
```diff
@@ -17,29 +17,30 @@ describe("instances.ScreenGui", function()
 			local instance = Instance.new("ScreenGui")
-			assert.equals(typeof(instance.AbsoluteSize), "Vector2")
+			assert.equals(type(instance.IgnoreGuiInset), "boolean")
 		end)
```
These all being separate `describe` tests is pretty strange, I don't remember when this happened.

ZoteTheMighty replied:
Yeah these can kind of all be consolidated into the above "should have properties..." test. I'll do that.

LPGhatguy commented on lib/instances/HttpService.lua:
```diff
@@ -11,4 +12,21 @@ function HttpService.prototype:JSONDecode(input)
+	if wrapInCurlyBraces == false then
+		return "04AEBFEA-87FC-480F-A98B-E5E221007A90"
+	else
+		return "{04AEBFEA-87FC-480F-A98B-E5E221007A90}"
```
I'll submit a followup change to generate real GUIDs; it's reasonable for a test to expect this function to work correctly.

LPGhatguy commented on lib/instances/UserInputService.lua:
```diff
@@ -1,3 +1,31 @@
+local UserInputService = BaseInstance:extend("UserInputService")
+
+function UserInputService.prototype:GetPlatform()
+	return Platform.Windows
```
🤔
